File: services/money.py
from __future__ import annotations

from decimal import ROUND_HALF_UP, Decimal

# Internal wallet balances and ledger amounts are stored as integer minor units
# of the rouble (kopecks). Product prices are quoted in whole roubles
# (`Tariff.price_rub`), so 1 RUB == 100 kopecks.
KOPECKS_IN_RUB = 100


def rubles_to_kopecks(rubles: int | str | Decimal) -> int:
    """Convert a rouble amount to integer kopecks (round half up)."""
    value = Decimal(str(rubles)) * KOPECKS_IN_RUB
    return int(value.quantize(Decimal("1"), rounding=ROUND_HALF_UP))


def kopecks_to_rubles(kopecks: int) -> Decimal:
    """Convert integer kopecks to a Decimal rouble amount."""
    return (Decimal(kopecks) / KOPECKS_IN_RUB).quantize(Decimal("0.01"))


def format_rub(kopecks: int) -> str:
    """Human-readable rouble string, dropping a trailing ``.00``."""
    amount = kopecks_to_rubles(kopecks)
    if amount == amount.to_integral_value():
        return f"{int(amount)}₽"
    return f"{amount.normalize()}₽"


def percent_of(amount_kopecks: int, percent: int) -> int:
    """Return ``percent`` percent of ``amount_kopecks`` (round half up)."""
    if percent < 0:
        raise ValueError("percent must be non-negative")
    value = Decimal(amount_kopecks) * Decimal(percent) / Decimal(100)
    return int(value.quantize(Decimal("1"), rounding=ROUND_HALF_UP))

File: web/src/lib/money.ts
/**
 * Money helpers — mirror of the backend `services/money.py`.
 * Internal amounts are integer kopecks (RUB minor units); 1 RUB = 100 kopecks.
 */
export const KOPECKS_IN_RUB = 100;

export function kopecksToRubles(kopecks: number): number {
  return kopecks / KOPECKS_IN_RUB;
}

/** Human-readable rouble string, dropping a trailing `.00`. */
export function formatRub(kopecks: number): string {
  const rub = kopecksToRubles(kopecks);
  const isWhole = Number.isInteger(rub);
  const body = isWhole
    ? rub.toLocaleString("ru-RU")
    : rub.toLocaleString("ru-RU", { minimumFractionDigits: 2, maximumFractionDigits: 2 });
  return `${body} ₽`;
}

/** Whole roubles → display string (for catalogue prices). */
export function formatRubFromRubles(rubles: number): string {
  return `${rubles.toLocaleString("ru-RU")} ₽`;
}

export function percentOf(amountKopecks: number, percent: number): number {
  return Math.round((amountKopecks * percent) / 100);
}
